pkg/rbd/controllerserver.go

```diff
@@ -265,93 +224,75 @@ func (cs *ControllerServer) DeleteVolume(ctx context.Context, req *csi.DeleteVol
-	if len(req.GetStartingToken()) > 0 {
-		i, parseErr := strconv.ParseUint(req.StartingToken, 10, 32)
-		if parseErr != nil {
-			return nil, status.Errorf(codes.Aborted, "invalid starting token %s", parseErr.Error())
-		}
-		//check starting Token is greater than list of rbd volumes
-		if len(rbdVolumes) < int(i) {
-			return nil, status.Errorf(codes.Aborted, "invalid starting token %s", parseErr.Error())
+		if _, ok := err.(ErrImageNotFound); !ok {
+			return nil, status.Error(codes.Internal, err.Error())
 		}
-		startToken = int(i)
-	}
 
-	var entries []*csi.ListVolumesResponse_Entry
+		// If image is missing, then there was a key to point to it, it means we need to cleanup
+		// the keys and return success
+		volumeNameMutex.LockKey(rbdVol.RequestName)
+		defer func() {
+			if err := volumeNameMutex.UnlockKey(rbdVol.RequestName); err != nil {
+				klog.Warningf("failed to unlock mutex volume:%s %v", rbdVol.RequestName, err)
+			}
+		}()
 
-	keys := make([]string, 0)
-	for k := range rbdVolumes {
-		keys = append(keys, k)
+		if err := unreserveVol(rbdVol, req.GetSecrets()); err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+		return &csi.DeleteVolumeResponse{}, nil
 	}
-	sort.Strings(keys)
 
-	for index, k := range keys {
-		if index < startToken {
-			continue
+	// lock out parallel create requests against the same volume name as we
+	// cleanup the image and associated omaps for the same
+	volumeNameMutex.LockKey(rbdVol.RequestName)
+	defer func() {
+		if err := volumeNameMutex.UnlockKey(rbdVol.RequestName); err != nil {
+			klog.Warningf("failed to unlock mutex volume:%s %v", rbdVol.RequestName, err)
 		}
-		entries = append(entries, &csi.ListVolumesResponse_Entry{
-			Volume: &csi.Volume{
-				VolumeId:      rbdVolumes[k].VolID,
-				CapacityBytes: rbdVolumes[k].VolSize,
-				VolumeContext: extractStoredVolOpt(rbdVolumes[k]),
-			},
-		})
-	}
+	}()
 
-	resp := &csi.ListVolumesResponse{
-		Entries: entries,
+	// Deleting rbd image
+	klog.V(4).Infof("deleting image %s", rbdVol.RbdImageName)
+	if err := deleteImage(rbdVol, rbdVol.AdminID, req.GetSecrets()); err != nil {
+		// TODO: can we detect "already deleted" situations here and proceed?
+		klog.V(3).Infof("failed to delete rbd image: %s/%s with error: %v",
+			rbdVol.Pool, rbdVol.RbdImageName, err)
+		return nil, status.Error(codes.Internal, err.Error())
 	}
 
-	return resp, nil
+	return &csi.DeleteVolumeResponse{}, nil
 }
 
 // ValidateVolumeCapabilities checks whether the volume capabilities requested
 // are supported.
 func (cs *ControllerServer) ValidateVolumeCapabilities(ctx context.Context, req *csi.ValidateVolumeCapabilitiesRequest) (*csi.ValidateVolumeCapabilitiesResponse, error) {
+	if req.GetVolumeId() == "" {
+		return nil, status.Error(codes.InvalidArgument, "Empty volume ID in request")
+	}
+
+	if len(req.VolumeCapabilities) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Empty volume capabilities in request")
```

Comment:
s/Empty/empty

Reply:
Discussion moved to [this](https://github.com/ceph/ceph-csi/pull/312#discussion_r282535932) comment for the issue pointed out.